cli/lib/kontena/cli/services/exec_command.rb

```diff
@@ -69,16 +47,116 @@ def execute
-        if !(container = service_containers.find{|container| container['instance_number'] == instance})
+        if !(container = service_containers.find{|c| c['instance_number'] == instance})
           exit_with_error "Service #{name} does not have container instance #{instance}"
         elsif container['status'] != 'running'
           exit_with_error "Service #{name} container #{container['name']} is not running, it is #{container['status']}"
-        else
+        elsif interactive?
+          interactive_exec(container)
+        else 
           exec_container(container)
         end
       else
-        exec_container(running_containers.first)
+        if interactive?
+          interactive_exec(running_containers.first)
+        else 
+          exec_container(running_containers.first)
+        end
       end
     end
+
+    # Exits if exec returns with non-zero
+    # @param [Docker::Container] container
+    def exec_container(container)
+      exit_status = nil
+      if !silent? && (verbose? || all?)
+        spinner "Executing command on #{container['name']}" do
+          exit_status = normal_exec(container)
+
+          raise Kontena::Cli::SpinAbort if exit_status != 0
+        end
+      else
+        exit_status = normal_exec(container)
+      end
+
+      exit exit_status if exit_status != 0 && !skip?
+
+      return exit_status == 0
+    end
+
+    # @param [Docker::Container] container
+    # @return [Boolean]
+    def normal_exec(container)
+      base = self
+      cmd = JSON.dump({ cmd: cmd_list })
+      exit_status = nil
+      token = require_token
+      url = ws_url(container)
+      url << '?shell=true' if shell?
+      ws = connect(url, token)
+      ws.on :message do |msg|
+        data = base.parse_message(msg)
+        if data 
+          if data['exit']
+            exit_status = data['exit'].to_i
+          elsif data['stream'] == 'stdout'
+            $stdout << data['chunk']
+          else 
+            $stderr << data['chunk']
+          end
+        end
+      end
+      ws.on :open do
+        ws.send(cmd)
+      end
+      ws.on :close do |e|
+        exit_status = 1
+      end
+
+      sleep 0.01 until !exit_status.nil?
+
+      exit_status
+    end
+
+    # @param [Docker::Container] container
+    def interactive_exec(container)
+      require 'io/console'
+
+      token = require_token
+      cmd = JSON.dump({ cmd: cmd_list })
+      base = self
+      url = ws_url(container) << '?interactive=true'
+      url << '&shell=true' if shell?
+      ws = connect(url, token)
+      ws.on :message do |msg|
+        base.handle_message(msg)        
+      end
+      ws.on :open do
+        ws.send(cmd)
+      end
+      ws.on :close do |e|
+        exit 1
+      end
+      
+      stream_stdin_to_ws(ws).join
+    end
+
+    # @param [Hash] container
+    # @return [String]
+    def ws_url(container)
+      "#{require_current_master.url.sub('http', 'ws')}/v1/containers/#{container['id']}/exec"
```

Comment:
Fails if the current master URL has a trailing `/`, the resulting `GET //v1/...` results in HTTP 404